// botpress/lib/cli/search.js
'use strict';

var _prompt = require('prompt');

var _prompt2 = _interopRequireDefault(_prompt);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = (argument, options) => {
  const getKeywords = _bluebird2.default.method(argument => {
    if (typeof argument === 'string') {
      return argument;
    } else {
      const schema = {
        properties: {
          keyword: {
            description: _chalk2.default.white('keyword:'),
            pattern: /^[a-zA-Z0-9][a-zA-Z0-9-_\.]+$/,
            message: 'keyword must be letters, digits, dashes, underscores and dots.',
            required: true
          }
        }
      };

      _prompt2.default.message = '';
      _prompt2.default.delimiter = '';
      _prompt2.default.start();

      return _bluebird2.default.fromCallback(callback => {
        _prompt2.default.get(schema, callback);
      }).then(({ keyword }) => keyword);
    }
  });

  const printResult = keyword => {
    console.log('You are actualy looking for: ' + keyword);
  };

  getKeywords(argument).then(printResult);
};
